src/Duende.AccessTokenManagement.OpenIdConnect/DPoPProofTokenHandler.cs

```diff
@@ -40,23 +39,33 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
         await SetDPoPProofTokenAsync(request, cancellationToken).ConfigureAwait(false);
         var response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
 
+        // The authorization server might send us a new nonce in a successful
+        // request, indicating that we should use the new nonce in future.
         var dPoPNonce = response.GetDPoPNonce();
 
-        // retry if 401
-        if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized && response.IsDPoPError())
+        if (dPoPNonce != null)
         {
-            response.Dispose();
+            _logger.LogDebug("The authorization server has supplied a new nonce");
 
-            await SetDPoPProofTokenAsync(request, cancellationToken, dPoPNonce).ConfigureAwait(false);
-            return await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
-        }
-        else if (dPoPNonce != null)
-        {
             await _dPoPNonceStore.StoreNonceAsync(new DPoPNonceContext
             {
                 Url = request.GetDPoPUrl(),
                 Method = request.Method.ToString(),
             }, dPoPNonce);
+
+            // But the authorization server might also send a failure response, and expect us to retry 
+            if (response.StatusCode == System.Net.HttpStatusCode.BadRequest)
```

Comment:
Ok, I think checking for 400 and the presence of the header is enough to know:

https://datatracker.ietf.org/doc/html/rfc9449#name-providing-a-new-nonce-value

and

https://datatracker.ietf.org/doc/html/rfc9449#name-authorization-server-provid